### PR TITLE
Add Files to the Offer type

### PR DESCRIPTION
**Pull request by vu3rdd**
Aeson instances and some minimal hedgehog tests for the offer message.

### CODE REVIEW

File: src/MagicWormhole/Internal/FileTransfer.hs [hunk 1]
```diff
@@ -20,19 +20,28 @@ import Data.Aeson
...
   , withObject
   )
+import System.Posix.Types (FileOffset)
 
 -- | An offer made by a sender as part of the Magic Wormhole file transfer protocol.
 --
 -- Currently only supports sending simple text messages. A full version would
 -- also support sending files and directories.
-newtype Offer
+data Offer
```
Comment (exarkun) [inline src/MagicWormhole/Internal/FileTransfer.hs]: Hm!  I'm still learning about `newtype` and trying to understand when it's a good idea to use `newtype` and when it's a good idea to use `data`.  Could you help me understand the change in this case?

File: src/MagicWormhole/Internal/FileTransfer.hs [hunk 2]
```diff
@@ -20,19 +20,28 @@ import Data.Aeson
...
 instance FromJSON Offer where
   parseJSON = withObject "Offer" $ \obj -> do
     offer <- obj .: "offer"
-    Message <$> offer .: "message"
+    asum [ Message <$> offer .: "message"
+         , File
+           <$> ((offer .: "file") >>= (.: "filename"))
+           <*> (toEnum <$> ((offer .: "file") >>= (.: "filesize")))
+         ]
+
```
Comment (exarkun) [inline src/MagicWormhole/Internal/FileTransfer.hs]: I had some trouble resolving `asum`.  It doesn't seem to be imported.  And it's not in Prelude, right?  Then I wondered if it was in Protolude.  But the best I could manage was to find http://hackage.haskell.org/package/protolude-0.2.2/docs/doc-index-All.html which seems to say yes, it is in Protolude, but I was unable to find any more information about it.  So the best I could do was find `Data.Foldable.asum` - and maybe Protolude is just re-exporting that?

What am I missing?

From context I guess I see what's happening here, maybe.  either `Message ...` or `File ...` could succeed (possibly both if you had a malformed peer I suppose).  `Data.Foldable.asum` would then give back the first one that succeeds.  That sounds like the right thing to do here.

File: tests/Generator.hs [hunk 3]
```diff
@@ -97,3 +101,9 @@ mailboxMessages = MailboxMessage <$> sides <*> phases <*> Gen.maybe messageIDs <
 
 welcomeMessages :: MonadGen m => m WelcomeMessage
 welcomeMessages = WelcomeMessage <$> Gen.maybe (Gen.text (Range.linear 0 1024) Gen.unicode) <*> Gen.maybe (Gen.text (Range.linear 0 1024) Gen.unicode)
+
+offerMessages :: MonadGen m => m F.Offer
+offerMessages = Gen.choice
+  [ F.Message <$> (Gen.text (Range.linear 0 1024) Gen.unicode)
+  , F.File <$> toS <$> (Gen.text (Range.linear 0 100) Gen.alphaNum)  <*> (fromIntegral <$> (Gen.int (Range.linear 0 maxBound)))
```
Comment (exarkun) [inline tests/Generator.hs]: I wonder if there are any concrete size limits in practice or if the server implementation will just take messages up to what its memory can handle.  There probably _should_ be some size limits if there aren't...  Maybe an issue to file against the Python implementation?

File: src/MagicWormhole/Internal/FileTransfer.hs (hunk 1 above)
Comment (vu3rdd) [reply src/MagicWormhole/Internal/FileTransfer.hs]: `newtype` is like type aliases but with more type safety and zero cost (of wrapping/unwrapping that you might expect) provided the type has only one constructor with one field. In the original code, there was only possible constructor with just one field (i.e. `Message`), so `data` and `newtype` could be interchanged. However, we added the `File` constructor with two fields inside below, so we can no longer use `newtype`.

File: src/MagicWormhole/Internal/FileTransfer.hs (hunk 2 above)
Comment (vu3rdd) [reply src/MagicWormhole/Internal/FileTransfer.hs]: `Protolude` exposes `asum` from `Data.Foldable`. https://github.com/sdiehl/protolude/blob/master/src/Protolude.hs (it qualifies the module as X and then reexports `X` at the top).

File: tests/Generator.hs (hunk 3 above)
Comment (vu3rdd) [reply tests/Generator.hs]: Good point. I will try finding that out from the python code base. Thanks!

**Pull request merged**